README.md
# dogstatsd-elixir

A fork of [adamkittelson/dogstatsd-elixir](https://github.com/adamkittelson/dogstatsd-elixir) with improvements and bug fixes.

A client for DogStatsD, an extension of the StatsD metric server for Datadog.

It is implemented as GenServer thus can be fitted into a supervision tree.

[![Build Status](https://travis-ci.org/envato/dogstatsd-elixir.svg?branch=master)](https://travis-ci.org/envato/dogstatsd-elixir)
[![Coverage Status](https://coveralls.io/repos/envato/dogstatsd-elixir/badge.png?branch=master)](https://coveralls.io/r/envato/dogstatsd-elixir?branch=master)

## Quick Start Guide

First install the library:

  1. Add dogstatsd to your `mix.exs` dependencies:

      ```elixir
      def deps do
        [
          {:dogstatsd, git: "https://github.com/envato/dogstatsd-elixir.git", tag: "0.0.4"}
        ]
      end
      ```

  2. Add `:dogstatsd` to your application dependencies:

      ```elixir
      def application do
        [applications: [:dogstatsd]]
      end
      ```
Then start instrumenting your code:

``` elixir
# Require the dogstatsd module.
require DogStatsd

# Configure DogStatsd.
{:ok, statsd} = DogStatsd.new("localhost", 8125)
```

Alternatively it can be placed in Application supervision tree:

```elixir
defmodule LoremIpsum.Application do
  use Application

  def start(_type, _args) do
    import Supervisor.Spec

    children = [
      worker(DogStatsd, [host: "localhost", port: 8125])
    ]

    opts = [strategy: :one_for_one, name: LoremIpsum.Supervisor]
    Supervisor.start_link(children, opts)
  end
end
```

Below are some examples how it is used:

```elixir
# Increment a counter.
DogStatsd.increment(statsd, "page.views")

# Record a gauge 50% of the time.
DogStatsd.gauge(statsd, "users.online", 123, %{sample_rate: 0.5})

# Sample a histogram
DogStatsd.histogram(statsd, "file.upload.size", 1234)

# Time a block of code
DogStatsd.time(statsd, "page.render") do
  render_page('home.html')
end

# Send several metrics at the same time
# All metrics will be buffered and sent in one packet when the block completes
DogStatsd.batch(statsd, fn(s) ->
  s.increment(statsd, "page.views")
  s.gauge(statsd, "users.online", 123)
end)

# Tag a metric.
DogStatsd.histogram(statsd, "query.time", 10, %{tags: ["version:1"]})
```

You can also post events to your stream. You can tag them, set priority and even aggregate them with other events.

Aggregation in the stream is made on hostname/event_type/source_type/aggregation_key.

``` elixir
# Post a simple message
DogStatsd.event(statsd, "There might be a storm tomorrow", "A friend warned me earlier.")

# Cry for help
DogStatsd.event(statsd, "SO MUCH SNOW", "Started yesterday and it won't stop !!", %{alert_type: "error", tags: ["urgent", "endoftheworld"]})
```

## Configuration

ENV | Use | Format | Default
--- | --- | --- | ---
DD_AGENT_HOST | The datadog-agent host | String | `127.0.0.1`
DD_DOGSTATSD_PORT| The datadog-agent Dogstatds port | String | `8125`

## Credits

* Based on [adamkittelson/dogstatsd-elixir](https://github.com/adamkittelson/dogstatsd-elixir)
* dogstatsd-elixir is a port of the [Ruby DogStatsd client](https://github.com/DataDog/dogstatsd-ruby)


CHANGELOG.md
CHANGELOG
=========

## New feature!

## 0.0.5-dev

### Enhancements

* Replaced deprecated String.to_char_list/1 with String.to_charlist/1

### Bug Fixes

### Deprecations

### Breaking Changes

* Rename environment variables:
  - DD_AGENT_PORT to DD_DOGSTATSD_PORT
  - DD_AGENT_ADDR to DD_AGENT_HOST

## 0.0.1 to 0.0.4

The CHANGELOG can be found in [the master branch of damkittelson/dogstatsd-elixir](https://github.com/adamkittelson/dogstatsd-elixir/blob/master/CHANGELOG.md)



lib/dogstatsd.ex
defmodule DogStatsd do
  use GenServer
  use DogStatsd.Statsd

  @default_host "127.0.0.1"
  @default_port 8125

  def new do
    start_link(%{})
  end

  def new(host, port, opts \\ %{}) do
    config = opts
             |> Map.put_new(:host, host)
             |> Map.put_new(:port, port)
    start_link(config)
  end

  def start_link(config, options \\ []) do
    config = config
             |> Map.take([:host, :port, :namespace, :tags, :max_buffer_size])
             |> Map.put_new(:max_buffer_size, 50)
             |> Map.put_new(:buffer, [])

    GenServer.start_link(__MODULE__, config, options)
  end

  def max_buffer_size(dogstatsd) do
    GenServer.call(dogstatsd, :max_buffer_size) || 50
  end

  def max_buffer_size(dogstatsd, buffer_size) do
    GenServer.call(dogstatsd, {:set_max_buffer_size, buffer_size})
  end

  def namespace(dogstatsd) do
    GenServer.call(dogstatsd, :get_namespace)
  end

  def namespace(dogstatsd, namespace) do
    GenServer.call(dogstatsd, {:set_namespace, namespace})
  end

  def host(dogstatsd) do
    GenServer.call(dogstatsd, :get_host) || System.get_env("DD_AGENT_HOST") || @default_host
  end

  def host(dogstatsd, host) do
    GenServer.call(dogstatsd, {:set_host, host})
  end

  def port(dogstatsd) do
    case GenServer.call(dogstatsd, :get_port) || System.get_env("DD_DOGSTATSD_PORT") || @default_port do
      port when is_binary(port) ->
        String.to_integer(port)
      port ->
        port
    end
  end

  def port(dogstatsd, port) do
    GenServer.call(dogstatsd, {:set_port, port})
  end

  def tags(dogstatsd) do
    GenServer.call(dogstatsd, :get_tags) || []
  end

  def tags(dogstatsd, tags) do
    GenServer.call(dogstatsd, {:set_tags, tags})
  end

  def prefix(dogstatsd) do
    case GenServer.call(dogstatsd, :get_namespace) do
      nil ->
        nil
      namespace ->
        "#{namespace}."
    end
  end

  def add_to_buffer(dogstatsd, message) do
    GenServer.call(dogstatsd, {:add_to_buffer, message})
  end

  def flush_buffer(dogstatsd) do
    buffer = dogstatsd
             |> GenServer.call(:flush_buffer)
             |> Enum.join("\n")
    send_to_socket dogstatsd, buffer
  end


  ###################
  # Server Callbacks

  def init(config) do
    {:ok, config}
  end

  def handle_call(:max_buffer_size, _from, config) do
    {:reply, config[:max_buffer_size], config}
  end

  def handle_call({:set_max_buffer_size, buffer_size}, _from, config) do
    config = config
             |> Map.put(:max_buffer_size, buffer_size)

    {:reply, config[:max_buffer_size], config}
  end

  def handle_call(:get_namespace, _from, config) do
    {:reply, config[:namespace], config}
  end

  def handle_call({:set_namespace, nil}, _from, config) do
    config = config
             |> Map.put(:namespace, nil)

    {:reply, config[:namespace], config}
  end

  def handle_call({:set_namespace, namespace}, _from, config) do
    config = config
             |> Map.put(:namespace, namespace)

    {:reply, config[:namespace], config}
  end

  def handle_call(:get_host, _from, config) do
    {:reply, config[:host], config}
  end

  def handle_call({:set_host, host}, _from, config) do
    config = config
             |> Map.put(:host, host)

    {:reply, config[:host], config}
  end

  def handle_call(:get_port, _from, config) do
    {:reply, config[:port], config}
  end

  def handle_call({:set_port, port}, _from, config) do
    config = config
             |> Map.put(:port, port)

    {:reply, config[:port], config}
  end

  def handle_call(:get_tags, _from, config) do
    {:reply, config[:tags], config}
  end

  def handle_call({:set_tags, tags}, _from, config) do
    config = config
             |> Map.put(:tags, tags)

    {:reply, config[:tags], config}
  end

  def handle_call({:add_to_buffer, message}, _from, config) do
    config = update_in(config, [:buffer], &(&1 ++ [message]))

    {:reply, config[:buffer], config}
  end

  def handle_call(:flush_buffer, _from, config) do
    {:reply, config[:buffer], Map.put(config, :buffer, [])}
  end

end
